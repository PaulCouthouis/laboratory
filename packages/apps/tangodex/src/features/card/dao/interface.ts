import type { Either } from 'purify-ts'
import type { Observable } from 'rxjs'
import type { Card } from '../domain/card'

export interface CardDAO {
  getById: (id: Card['id']) => Observable<Either<string, Card>>
}
