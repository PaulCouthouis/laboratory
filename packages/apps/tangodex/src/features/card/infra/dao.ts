import type { Either } from 'purify-ts'

import { from, map, Observable, take } from 'rxjs'
import {
  retrieveRecordsFromWordCardById,
  toFirstFieldSet,
} from '../../../db/airtable'
import { Card, decodeCard } from '../core/domain/card'
import type { Pile } from '../core/domain/pile'

export interface CardDAO {
  getById: (id: Card['id']) => Observable<Either<string, Card>>
  getByIds: (ids: Card['id'][]) => Observable<Either<string, Pile>>
}

export const CardDAO: () => CardDAO = () => {
  const getById = (id: Card['id']) => {
    const records = retrieveRecordsFromWordCardById(id)

    return from(records).pipe(map(toFirstFieldSet), map(decodeCard), take(1))
  }

  return { getById }
}
