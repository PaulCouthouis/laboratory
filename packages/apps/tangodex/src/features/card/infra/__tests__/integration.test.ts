import { expect, test } from 'vitest'
import { CardDAO } from '../dao'
import { firstValueFrom } from 'rxjs'

test('Get One Card', async () => {
  const dao = CardDAO()

  const card$ = dao.getById('KDS-0001')
  const card = await firstValueFrom(card$)

  expect(card.extract()).toMatchSnapshot()
})

test('Get Card Pile', async () => {
  const dao = CardDAO()

  const cards$ = dao.getByIds(['KDS-0002', 'KDS-0003', 'KDS-0004'])
  const cards = await firstValueFrom(cards$)

  expect(cards.extract()).toMatchSnapshot()
})