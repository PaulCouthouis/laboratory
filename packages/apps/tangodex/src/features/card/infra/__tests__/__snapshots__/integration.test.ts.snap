// Vitest Snapshot v1

exports[`Get Card Pile 1`] = `
[
  {
    "category": "kandoushi",
    "description": "Il faut le connaitre mais éviter de l'utiliser car dire \\"non\\" en japonais peut paraitre très froid.

Utilisable cependant après que quelqu'un vous complimente pour le refuter et paraitre plus modeste.",
    "id": "KDS-0002",
    "illustrator": "??",
    "name": "いいえ",
    "sentences": "山本：「田中さんですか？」山田：「いいえ。」",
    "translation": "Non",
    "variant": undefined,
  },
  {
    "category": "kandoushi",
    "description": "Se dit uniquement le matin. L'ajout de ございます permet d'être plus formel",
    "id": "KDS-0003",
    "illustrator": "??",
    "name": "おはよう",
    "sentences": "先生！おはようございます。",
    "translation": "Bonjour (matin)",
    "variant": "おはようございます",
  },
  {
    "category": "kandoushi",
    "description": "Utilisable seulement l'après-midi",
    "id": "KDS-0004",
    "illustrator": "??",
    "name": "こんにちは",
    "sentences": "こんにちは！もう食べましたか？",
    "translation": "Bonjour (après-midi)",
    "variant": undefined,
  },
]
`;

exports[`Get One Card 1`] = `
{
  "category": "kandoushi",
  "description": "はい est la version la plus fréquente.
Il existe aussi ええ qui doit être cependant suivi par une phrase.",
  "id": "KDS-0001",
  "illustrator": "stampo.fun",
  "name": "はい",
  "sentences": "山本：「田中さんですか？」田中：「はい。」
山本：「田中さんですか？」田中：「ええ、そうです。」",
  "translation": "Oui",
  "variant": "ええ",
}
`;
